src/SemiFungible1155.sol

```diff
@@ -14,27 +15,43 @@ error NotAllowed();
 error TypeMismatch();
 error FractionalBurn();
 
+/// @title Contract for minting semi-fungible EIP1155 tokens
+/// @author bitbeckers
+/// @notice Extends { Upgradeable1155 } token with semi-fungible properties and the concept of `units`
+/// @dev Adds split bit strategy as described in [EIP-1155](https://eips.ethereum.org/EIPS/eip-1155#non-fungible-tokens)
 contract SemiFungible1155 is Upgradeable1155 {
+    /// @dev Counter used to generate next typeID.
     uint256 public typeCounter;
-    // Use a split bit implementation.
-    // Store the type in the upper 128 bits..
+
+    /// @dev Bitmask used to expose only upper 128 bits of uint256
     uint256 public constant TYPE_MASK = uint256(uint128(int128(~0))) << 128;
 
-    // ..and the non-fungible index in the lower 128
+    /// @dev Bitmask used to expose only lower 128 bits of uint256
     uint256 public constant NF_INDEX_MASK = uint128(int128(~0));
 
-    // The top bit is a flag to tell if this is a NFI.
+    /// @dev Identify non-fungible index. Use to find index of token belonging to `typeID`
     uint256 public constant TYPE_NF_BIT = uint256(1 << 255);
 
+    /// @dev Mapping of `tokenID` to address of `owner`
     mapping(uint256 => address) internal owners;
+
+    /// @dev Mapping of `tokenID` to address of `creator`
     mapping(uint256 => address) internal creators; //TODO extend with admin contracts
 
+    /// @dev Used to determine amount of `units` stored in token at `tokenID`
     mapping(uint256 => uint256) internal tokenValues;
+
+    /// @dev Used to find highest index of token belonging to token at `typeID`
+    // TODO should have max value type(uint256).max
     mapping(uint256 => uint256) internal maxIndex;
+
+    /// @dev Mapping from `tokenID` to user at `address` to get `units` owned
     mapping(uint256 => mapping(address => uint256)) internal tokenUserBalances;
```

Comment:
Is this data structure necessary?

My understanding is that a tokenID represents a fractional ownership of a hypercert correct?
If so, then only 1 address could possibly own that fraction.
So we should be able to easily reconstruct what. we need from `owners` and `tokenValues`?

Does storing this help reduce gas costs in some way that I'm not seeing?

Reply:
No, it might even be way more efficient because we reduce a lot of state changes. 

If the `tokenID` is of a base type, it should be zero because that's just an entry to hold state and data. The base type itself hold the totalValue but that would not be owned by an account.

Something like this:
```solidity
    function _unitsOf(address account, uint256 tokenID) internal view returns (uint256 units) {
        units = 0;

        // Check if fraction token and accounts owns it
        if (getNonFungibleIndex(tokenID) != 0 && ownerOf(tokenID) == account) {
            units = tokenValues[tokenID];
        }

        return units;
    }
 ```

Reply:
That makes sense to me. I think calculating total per user can easily be done off-chain by just adding up all my NFT units right?

Reply:
Created an issue for it so we can resolve this thread